src/libertem/common/progress.py

```diff
@@ -0,0 +1,343 @@
+                    ...
+            except WorkerQueueEmpty:
+                break
+
+    def monitor_queue(self):
+        """
+        Monitor the queue for messages
+        If there are no subscribers this should drain
+        messages from the queue as fast as they are recieved
+        """
+        while True:
+            with self._message_q.get(block=True) as ((topic, msg), _):
+                if topic == 'STOP':
+                    break
+                try:
+                    for callback in self._subscriptions[topic]:
+                        callback(topic, msg)
+                except KeyError:
+                    pass
+
+
+class ProgressManager:
+    """
+    Handle construction and updating of a tqdm.tqdm progress bar
+    for a set of UDFTasks, to be completed in any order
+
+    The bar displays as such:
+
+        Partitions: n_complete(n_in_progress) / n_total, ...\
+            Frames: [XXXXX..] frames_completed / total_frames ...
+
+    When processing tile stacks, stacks are treated as frames
+    as such: (pseudo_frames = tile.size // sig_size)
+
+    The bar will render in a Jupyter notebook as a JS widget
+    automcatially via tqdm.auto
+    """
+    def __init__(self, tasks: Iterable['UDFTask']):
+        if not tasks:
+            raise ValueError('Cannot display progress for empty tasks')
+        # the number of whole frames we expect each task to process
+        self._task_max = {t.partition.get_ident(): t.task_frames
+                          for t in tasks}
+        # _counters is our record of progress on a task,
+        # values are floating whole frames processed
+        # as in tile mode we can process part of a frame
+        self._counters = {k: 0. for k in self._task_max.keys()}
+        total_frames = sum(self._task_max.values())
+        # For converting tiles to pseudo-frames
+        self._sig_size = tasks[0].partition.shape.sig.size
+        # Counters for part display
+        self._num_complete = 0
+        self._num_in_progress = 0
+        self._num_total = len(self._counters)
+        # Create the bar object
+        self._bar = self.make_bar(total_frames)
+
+    def make_bar(self, maxval: int):
+        from tqdm.auto import tqdm
+        return tqdm(desc=self.get_description(),
+                    total=maxval,
+                    leave=True)
+
+    def get_description(self) -> str:
+        """
+        Get the most recent description string for the
+        bar, including partition information
+        If we know that partitions are in progress
+        include this in parentheses after n_completed
+
+        Only called when we start/end a partition
+        """
+        if self._num_in_progress:
+            return (f'Partitions {self._num_complete}({self._num_in_progress})'
+                    f'/{self._num_total}, Frames')
+        else:
+            return (f'Partitions {self._num_complete}'
+                    f'/{self._num_total}, Frames')
```

Comment:
This could also be interesting for testing - inject a different `ProgressReporter` into the `ProgressManager`, which just keeps a history of `start`/`end`/`update` calls, and see if that history matches what we expect. I guess a complete integration test could be more difficult, as we can't really precisely control the partition timings etc.

But maybe we should at least add a test to the executor integration suite that sets `progress=True`? That we make sure that 1) we don't crash, and 2) we don't hang at the end, when we enable progress with the different executors.

Another strategy could be to have a UDF that is written in a way that at least one partition is "slow" to process, and that we re-define slow for this case to be "takes longer than 100ms" using `min_message_interval`. Then we can check our mock progress reporter, which should contain at least one progress message. Could that work?

Reply:
All good advice, thanks. I've abstracted away the `tqdm` part, and am just looking at where to insert the choice of reporter. One option is right at the top at `lt.Context()`, the reasoning being that it serves a similar role to the `plot_class` kwarg that is already there. I don't really like this though as I can't see many cases where this needs to be user-facing!

At the moment I don't know enough about the `libertem.web` architecture to know where is best to put it, so for now I'll make it an optional argument on the standard `UDFRunner` (non-user-facing) and we can sort it later ?

Reply:
All done and used for testing. In the tests I inject a subclass of the normal TQDM reporter which keeps a copy of every update it is given, then I use the history to check we recieved a set of messages which are plausible. This is coupled to a UDF which processes one tile very slowly, and another test where I mock time.time() to count in 1.5 second increments so we get lots of messages.